lib/Dialect/ESI/cosim/cosim_dpi_server/Endpoint.cpp

```diff
@@ -0,0 +1,70 @@
+//===- EndPoint.cpp - Definitions for EndPointRegistry ----------*- C++ -*-===//
+//
+// Definitions for Cosim EndPoint and EndPointRegistry.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/cosim/Endpoint.h"
+
+using namespace circt::esi::cosim;
+
+Endpoint::Endpoint(uint64_t sendTypeId, int sendTypeMaxSize,
+                   uint64_t recvTypeId, int recvTypeMaxSize)
+    : sendTypeId(sendTypeId), recvTypeId(recvTypeId), inUse(false) {}
+Endpoint::~Endpoint() {}
+
+uint64_t Endpoint::getSendTypeId() const { return sendTypeId; }
+uint64_t Endpoint::getRecvTypeId() const { return recvTypeId; }
+
+bool Endpoint::setInUse() {
+  Lock g(m);
+  if (inUse)
+    return false;
+  inUse = true;
+  return true;
+}
+
+void Endpoint::returnForUse() {
+  Lock g(m);
+  if (!inUse)
+    throw std::runtime_error("Cannot return something not in use!");
+  inUse = false;
+}
+
+void EndpointRegistry::registerEndpoint(int epId, uint64_t sendTypeId,
+                                        int sendTypeMaxSize,
+                                        uint64_t recvTypeId,
+                                        int recvTypeMaxSize) {
+  Lock g(m);
+  if (endpoints.find(epId) != endpoints.end())
+    throw std::runtime_error("Endpoint ID already exists!");
+  endpoints.emplace(std::piecewise_construct, std::forward_as_tuple(epId),
```

Comment:
All the metaprogramming here with piecewise_construct etc seems like overkill for a tuple of primitive types.

Reply:
I agree. The problem is that I'm trying to insert an entry with a value class from which I've deleted the copy constructor into a map. This is the only way I could figure to do it, though my metaprogramming fu is weak. I've added comments to that effect.